exec/memory/base/src/main/java/org/apache/drill/exec/memory/BaseAllocator.java

```diff
@@ -0,0 +1,689 @@
+  }
+
+  /**
+   * Return a unique Id for an allocator. Id's may be recycled after a long period of time.
+   *
+   * <p>
+   * Primary use for this is for debugging output.
+   * </p>
+   *
+   * @return the allocator's id
+   */
+  long getId() {
+    return id;
+  }
+
+  @Override
+  public BufferAllocator newChildAllocator(
+      final String name,
+      final long initReservation,
+      final long maxAllocation) {
+    final ChildAllocator childAllocator = new ChildAllocator(this, name, initReservation, maxAllocation);
+
+    if (DEBUG) {
+      childAllocators.put(childAllocator, childAllocator);
+      historicalLog.recordEvent("allocator[%d] created new child allocator[%d]",
+          id, childAllocator.id);
+    }
+
+    return childAllocator;
+  }
+
+  private class Reservation extends AllocationReservation {
+    private final HistoricalLog historicalLog;
+
+    public Reservation() {
+      if (DEBUG) {
+        historicalLog = new HistoricalLog("Reservation[allocator[%d], %d]", id, System.identityHashCode(this));
+        historicalLog.recordEvent("created");
+        synchronized (DEBUG_LOCK) {
+          reservations.put(this, this);
+        }
+      } else {
+        historicalLog = null;
+      }
+    }
+
+    @Override
+    public void close() {
+      if (DEBUG) {
+        if (!isClosed()) {
+          final Object object;
+          synchronized (DEBUG_LOCK) {
+            object = reservations.remove(this);
+          }
+          if (object == null) {
+            final StringBuilder sb = new StringBuilder();
+            print(sb, 0, Verbosity.LOG_WITH_STACKTRACE);
+            logger.debug(sb.toString());
+            throw new IllegalStateException(
+                String.format("Didn't find closing reservation[%d]", System.identityHashCode(this)));
+          }
+
+          historicalLog.recordEvent("closed");
+        }
+      }
+
+      super.close();
+    }
+
+    @Override
+    protected boolean reserve(int nBytes) {
+      final AllocationOutcome outcome = BaseAllocator.this.allocateBytes(nBytes);
+
+      if (DEBUG) {
+        historicalLog.recordEvent("reserve(%d) => %s", nBytes, Boolean.toString(outcome.isOk()));
+      }
+
+      return outcome.isOk();
+    }
+
+    @Override
+    protected DrillBuf allocate(int nBytes) {
+      boolean success = false;
+
+      /*
+       * The reservation already added the requested bytes to the allocators owned and allocated bytes via reserve().
+       * This ensures that they can't go away. But when we ask for the buffer here, that will add to the allocated bytes
+       * as well, so we need to return the same number back to avoid double-counting them.
+       */
+      try {
+        final DrillBuf drillBuf = BaseAllocator.this.bufferWithoutReservation(nBytes, null);
+
+        if (DEBUG) {
+          historicalLog.recordEvent("allocate() => %s",
+              drillBuf == null ? "null" : String.format("DrillBuf[%d]", drillBuf.getId()));
```

Comment:
`drillBuf == null` check is unnecessary, `bufferWithoutReservation()` cannot return `null`